src/loaders.py

```diff
@@ -9,52 +11,80 @@
+        self.sample_size = len(self.images_data) if sample_size is None else min(sample_size, len(self.images_data))
 
     def __iter__(self):
-        return iter(torch.randperm(self.data_size)[:self.sample_size].long())
+        indices = random.sample(range(len(self.images_data)), self.sample_size)
+        return iter(indices)
+
+    def __len__(self):
+        return self.sample_size
+
+
+class AspectRatioSampler(Sampler):
+    def __init__(self, images_data, *, batch_size=8, sample_size=None):
+        self.images_data = images_data
+        self.batch_size = batch_size
+        self.sample_size = len(self.images_data) if sample_size is None else min(sample_size, len(self.images_data))
+        self.sample_size = self.sample_size // self.batch_size * self.batch_size
+
+        self.indices = self.images_data.sort_values('aspect_ratio').index.tolist()
+
+    def __iter__(self):
+        subset = sorted(random.sample(range(len(self.indices)), self.sample_size))
+        indices = np.array([self.indices[i] for i in subset])\
+            .reshape(self.sample_size // self.batch_size, self.batch_size)
+        np.random.seed()
+        np.random.shuffle(indices)
+        indices = indices.flatten()
+
+        return iter(indices)
 
     def __len__(self):
         return self.sample_size
 
 
 class ImageDetectionDataset(Dataset):
-    def __init__(self, ids, annotations, annotations_human_labels, images_dir, target_encoder, train_mode,
-                 image_transform=None):
+    def __init__(self, images_data, annotations, annotations_human_labels, target_encoder, train_mode,
+                 short_dim, long_dim, image_h, image_w, sampler_name, image_transform):
         super().__init__()
-        self.ids = ids
+        self.images_data = images_data
         self.annotations = annotations
         self.annotations_human_labels = annotations_human_labels
-        self.images_dir = images_dir
         self.target_encoder = target_encoder
         self.train_mode = train_mode
 
+        self.short_dim = short_dim
+        self.long_dim = long_dim
+        self.image_h = image_h
+        self.image_w = image_w
+        self.sampler_name = sampler_name
         self.image_transform = image_transform
 
     def __len__(self):
-        return len(self.ids)
+        return len(self.images_data)
 
     def __getitem__(self, index):
-        Xi = self.load_from_disk(index)
 
-        Xi = self.image_transform(Xi)
+        Xi = self.load_from_disk(index)
+        Xi = self.resize_image(Xi)
 
         if self.annotations is not None:
-            _, h, w = Xi.size()
+            w, h = Xi.size
             yi = self.load_target(index, (h, w))
             return Xi, yi
         else:
-            return Xi
+            return self.image_transform(Xi)
 
     def load_from_disk(self, index):
-        imgId = self.ids[index]
-        img_path = os.path.join(self.images_dir, imgId + '.jpg')
+        img_path = self.images_data.iloc[index]['image_path']
```

Comment:
@i008 ok, I think I agree, in metadata we would have all the info about image, but we create paths on the fly so that we can use the same meta file in every env, do you agree @jakubczakon?

Reply:
yeah that makes sense. Regenerating it would be sad I think